src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.cs

```diff
@@ -0,0 +1,267 @@
+            RuleId,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeTitle)),
+            CreateLocalizableResourceString(nameof(UseConcreteTypeForMethodReturnMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        internal static readonly DiagnosticDescriptor UseConcreteTypeForParameter = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeTitle)),
+            CreateLocalizableResourceString(nameof(UseConcreteTypeForParameterMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        internal static readonly DiagnosticDescriptor UseConcreteTypeForLocal = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeTitle)),
+            CreateLocalizableResourceString(nameof(UseConcreteTypeForLocalMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        internal static readonly DiagnosticDescriptor UseConcreteTypeForField = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeTitle)),
+            CreateLocalizableResourceString(nameof(UseConcreteTypeForFieldMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(UseConcreteTypeForField);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            if (!Debugger.IsAttached)
+            {
+                context.EnableConcurrentExecution();
+            }
+
+            context.RegisterCompilationStartAction(c =>
+            {
+                var state = new State(c.Compilation);
+                var coll = new Collector(state);
+
+                c.RegisterOperationAction(ctx => coll.HandleInvocation((IInvocationOperation)ctx.Operation), OperationKind.Invocation);
+                c.RegisterOperationAction(ctx => coll.HandleSimpleAssignment((ISimpleAssignmentOperation)ctx.Operation), OperationKind.SimpleAssignment);
```

Comment:
Do we handle other assignment kinds such as compound assignment and coalesce assignment: https://sourceroslyn.io/#Microsoft.CodeAnalysis/Generated/Operations.Generated.cs,30dfd6516f6a7ad0,references?

Also wanted to confirm this will handle passing a local/parameter/field as an out/ref parameter to method calls. I'll check the tests to confirm anyways.

Reply:
This was originally an FxCop rule I wrote back in ~2009 or so, coalescing operations weren't a thing back then. So I missed those cases when porting this to Roslyn.

I added support for coalescing assignments and expressions. Compound assignments don't make sense in this context since they only apply to non-upgradable types.

I do have support for ref/out in there already.